### ISSUE TITLE
Fluent API instead of MapValue attribute

### ISSUE THREAD

**Issue by boris-turk** [linq2db/linq2db#2576]
Hi,

is it possible to define enum mappings WITHOUT using attributes, maybe with some kind of fluent API? The only possibility I see right now is to use the `MapValue` attribute:
```cs
	public enum Gender
	{
		[MapValue("M")] Male,
		[MapValue("F")] Female,
		[MapValue("U")] Unknown,
		[MapValue("O")] Other,
	}
```
But I don't like this approach (see my longer explanation below). Thank you in advance for any kind of help! Best regards,

Boris Turk

Our team is developing a bigger ERP system and we decided to use the onion architecture (together with dependency injection):

https://jeffreypalermo.com/2008/07/the-onion-architecture-part-1/

The inner part of our onion architecture are so-called domain objects. These are more or less just simple DTO classes (like Article, Address, Service) and some common enums, e.g.:
```cs
	public enum DocumentPostingFlag
	{
		[StorageValue(" ")]
		None,
		
		[StorageValue("O")]
		Open,
		
		[StorageValue("V")]
		Posted,
	}
```
`StorageValue` is our own attribute, basically the same thing as `MapValue`. And here comes the problem: we would like to keep our core assemblies (like domain objects) independent of linq2db assembly. This kind of architectural rule looks quite restrictive at first glance, but it pays off in the long term. Most of our domain logic stays decoupled from a specific persistence technology. One day we might be forced to switch to another ORM provider and it will be easier to do that if we don't make the majority of our system dependent on third-party code.

We were initially playing with the idea that we let our core assemblies depend on linq2db, but ensure that we keep the dependencies at minimum - i.e. only use the `MapValue` attributes and keep the rest of linq2db initialization (fluent API rules) in one of the marginal assemblies, at the boundary of onion architecture. But I don't like this approach as it requires us to constantly monitor that the dependencies did not go out of control. This is hard to ensure in a team of 10 developers.

Let me point out that we were already able to isolate the majority of linq2db specifics into its own assembly which is defined at the boundary of our architecture. The only part of the puzzle that is still causing problems is `MapValue`.

**Comment by MaceWindu** [comment]
Currently there is no enum-specific fluent mapping API, but this should work:
```cs
mb.HasAttribute(typeof(SomeEnum).GetField(nameof(SomeEnum.Value1)), new MapValueAttribute(100));
mb.HasAttribute(typeof(SomeEnum).GetField(nameof(SomeEnum.Value2)), new MapValueAttribute(200));
```

**Comment by boris-turk** [comment]
Thanks a lot @MaceWindu for the quick reply. I tested it and it works like a charm! I have to say it once again: you are all doing a great job! It must be quite a challenge to build up a product like linq2db - and you are doing it for free! Respect :)

**Comment by feeleen** [comment]
Why we cant use custom TypeConverters instead of MapValue attributes?

For example if i define my own TypeConverter for my class or EnumConverter for my enum, then i don't depend on linq2db map attributes

Moreover if my enum contains huge amount of values, i must map each with MapValue attribute, painful

Recently I met problems with default linq2db converter (it uses simple Convert.ChangeType(..)) and not respecting my custom type converters, defined on user classes).

```cs
	[TypeConverter(typeof(StringEnumConverter<PersonType>))]
	public class PersonType : StringEnum<PersonType>
	{
		public static PersonType EM => New();
		public static PersonType SP => New();
		public static PersonType SC => New();
		public static PersonType VC => New();
		public static PersonType IN => New();
	}

	[Table(Name = "[Person].[Person]")]
	public class Person
	{
		[PrimaryKey, Identity]
		public int BusinessEntityID { get; set; }
		[Column]
		public PersonType PersonType { get; set; }
	}
```
as far as i understand this code doesn't work since we cant use ChangeType to convert strings to custom classes, even if there are implicit operators and type converters exists.

It would be great if type convertion in linq2db was made regarding custom TypeConverters

maybe it's worh a new issue, not sure..

**Comment by feeleen** [comment]
I don't know all nuances of ConvertBuilder class, it's too complicated logic for me at first sight,
but when I made a quick test, just added to ConvertBuilder.ConvertDefault method some TypeConverter discovering logic:

```cs

static object ConvertDefault(object value, Type conversionType)
{
	try
	{
		// new logic starts here
		if (!conversionType.IsPrimitive && conversionType != typeof(string) && conversionType != typeof(Data.DataParameter))
		{
			System.ComponentModel.TypeConverter typeConverter = System.ComponentModel.TypeDescriptor.GetConverter(conversionType);
			if (typeConverter != null)
				return typeConverter.ConvertFrom(null, Thread.CurrentThread.CurrentCulture, value);
		}
		// ends here
		return Convert.ChangeType(value, conversionType, Thread.CurrentThread.CurrentCulture);
	}
	catch (Exception ex)
	{
		throw new LinqToDBConvertException($"Cannot convert value '{value}: {value.GetType().FullName}' to type '{conversionType.FullName}'", ex);
	}
}

```

it worked for my test above like a charm and I got rid of MapValue attribute

**Comment by sdanyliv** [comment]
@feeleen, have you tried `HasConversion` when defining entity via `FluentMappingBuilder`? You can define converter for specific property.

**Comment by feeleen** [comment]
@sdanyliv 

tried few things now with no luck:

```cs
var ms = new MappingSchema();
var builder = ms.GetFluentMappingBuilder();

_ = builder.Entity<Person>().Property(e => e.PersonType).HasConversion(v => v.Value, s => PersonType.Parse(s));

ms.SetConverter<string, PersonType>(s => PersonType.Parse(s));
ms.SetConverter<PersonType, DataParameter>(val => new DataParameter { Value = val.Value, DataType = DataType.VarChar });

ms.SetValueToSqlConverter(typeof(PersonType), (sb, tp, v) =>
{
    tp = LinqToDB.SqlQuery.SqlDataType.DbVarChar;
    _ = sb.Append('\'').Append((string)v).Append('\'');  // -> funny things happen here )))) don't know really what to write here, but it didn't help anyway
});
```

error is 

```cs
Message: 
    Test method TestProject1.UnitTest1.TestMethod1 threw exception: 
    LinqToDB.Common.LinqToDBConvertException: Cannot convert value 'EM: System.String' to type 'TestProject1.PersonType' ---> System.InvalidCastException: Invalid cast from 'System.String' to 'TestProject1.PersonType'.

  Stack Trace: 
    Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
    IConvertible.ToType(Type type, IFormatProvider provider)
    Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
    ConvertBuilder.ConvertDefault(Object value, Type conversionType) line 29
    --- End of inner exception stack trace ---
    ConvertBuilder.ConvertDefault(Object value, Type conversionType) line 33
    lambda_method176(Closure , IDataReader )
    ColumnReader.GetValue(IDataReader dataReader) line 304
    lambda_method174(Closure , IQueryRunner , IDataReader )
    Mapper`1.Map(IDataContext context, IQueryRunner queryRunner, IDataReader dataReader) line 117
    QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 429
    QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 432
    QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 433
    ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken) line 183
    AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token) line 203
    UnitTest1.GetPersonsTableAsync() line 19
    UnitTest1.TestMethod1() line 30
    ThreadOperations.ExecuteWithAbortSafety(Action action)

```

**Comment by sdanyliv** [comment]
It should be enough to define just
```cs
_ = builder.Entity<Person>().Property(e => e.PersonType).HasConversion(v => v.Value, s => PersonType.Parse(s));
```
If it is not working, I need model and query when it fails.

**Comment by feeleen** [comment]
One more thing i should notice is that

```cs
ms.SetConverter<PersonType, DataParameter>(val => new DataParameter { Value = val.Value, DataType = DataType.VarChar });
```

didn't help me tu update record

only one thing helped is to put DataType attribute on Entity property

```cs
[Table(Name = "[Person].[Person]")]
public class Person
{
    [PrimaryKey, Identity]
    public int BusinessEntityID { get; set; }
    [Column]
    [DataType(DataType.VarChar)] // <-- this helped only
    public PersonType PersonType { get; set; }
}
```

**Comment by feeleen** [comment]
My test model for AdventureWorks2017 database is:

```cs
public class DbContext : DataConnection
{
    public DbContext() : base("SqlServer", @"data source=.\SQLEXPRESS01;
			     initial catalog=AdventureWorks2017;persist security info=True; 
			     Integrated Security=SSPI;")
    {
    }
    
    [TypeConverter(typeof(StringEnumConverter<PersonType>))]
    public class PersonType : StringEnum<PersonType>
    {
        public static PersonType EM => New();
        public static PersonType SP => New();
        public static PersonType SC => New();
        public static PersonType VC => New();
        public static PersonType IN => New();
        public static PersonType GC => New();
    }
    
    [Table(Name = "[Person].[Person]")]
    public class Person
    {
        [PrimaryKey, Identity]
        public int BusinessEntityID { get; set; }
        [Column]
        [DataType(DataType.VarChar)]
        public PersonType PersonType { get; set; }
        [Column]
        public int NameStyle { get; set; }
        [Column]
        public string Title { get; set; }
        [Column]
        public string FirstName { get; set; }
        [Column]
        public string MiddleName { get; set; }
        [Column]
        public string LastName { get; set; }
        [Column]
        public string Suffix { get; set; }
        [Column]
        public bool EmailPromotion { get; set; }
        [Column]
        public string AdditionalContactInfo { get; set; }
        [Column]
        public string Demographics { get; set; }
        [Column]
        public Guid Rowguid { get; set; }
        [Column]
        public DateTime? ModifiedDate { get; set; }
    }
}
```

generated sql is
```sql
SELECT
	[t1].[BusinessEntityID],
	[t1].[PersonType],
	[t1].[NameStyle],
	[t1].[Title],
	[t1].[FirstName],
	[t1].[MiddleName],
	[t1].[LastName],
	[t1].[Suffix],
	[t1].[EmailPromotion],
	[t1].[AdditionalContactInfo],
	[t1].[Demographics],
	[t1].[Rowguid],
	[t1].[ModifiedDate]
FROM
	[Person].[Person] [t1]
```

Test code
```cs
using (var db = new DbContext())
{
    DbContext.InitContextMappings();
    
    var persons = await db.GetTable<Person>().ToListAsync();
    
    var res = db.Update(persons[0], (a, b) => b.ColumnName == "PersonType");
}
```

**Comment by sdanyliv** [comment]
While check this tomorrow,  thanks.

**Comment by feeleen** [comment]
Also tried with simple enum 

```cs
public enum MyPersonType
{
    EM,
    SP,
    SC,
    VC,
    IN,
    GC
}
```
with
```cs
_ = builder.Entity<Person>().Property(e => e.PersonType).HasConversion(v => v.ToString(), s => Enum.Parse<MyPersonType>(s));
			ms.SetConverter<string, MyPersonType>(s => Enum.Parse<MyPersonType>(s));
			ms.SetConverter<MyPersonType, DataParameter>(val => new DataParameter { Value = val.ToString(), DataType = DataType.VarChar });

```

it didn't not work

```
Message: 
Test method TestProject1.UnitTest1.TestMethod1 threw exception: 
LinqToDB.Common.LinqToDBConvertException: Mapping of column 'PersonType' value failed, see inner exception for details ---> System.FormatException: Input string was not in a correct format.

  Stack Trace: 
Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
Number.ParseInt32(ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
lambda_method139(Closure , IDataReader )
ColumnReader.GetValue(IDataReader dataReader) line 304
--- End of inner exception stack trace ---
ColumnReader.GetValue(IDataReader dataReader) line 314
lambda_method136(Closure , IQueryRunner , IDataReader )
Mapper`1.Map(IDataContext context, IQueryRunner queryRunner, IDataReader dataReader) line 117
QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 429
QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 432
QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken) line 433
ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken) line 183
AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token) line 203
UnitTest1.GetPersonsTableAsync() line 19
UnitTest1.TestMethod1() line 30
ThreadOperations.ExecuteWithAbortSafety(Action action)
```

but when I added MapValue attribute on enum members all worked fine...

**Comment by sdanyliv** [comment]
Yes, you have to define Column DataType, or:
```cs
ms.SetDataType(typeof(MyPersonType), DataType.VarChar);
```

**Comment by feeleen** [comment]
Made some additional test and discovered that 

SetConverter via DataConnection's member MappingSchema doesn't work:

```cs
public class DbContext : DataConnection
{
    public DbContext() : base("SqlServer", @"data source=.\SQLEXPRESS01;
			     initial catalog=AdventureWorks2017;persist security info=True; 
			     Integrated Security=SSPI;")
    {
    }

    public static void InitContextMappings()
    {
	    this.MappingSchema.SetConverter<string, MyPersonType>(s => Enum.Parse<MyPersonType>(s));
	    this.MappingSchema.SetConverter<MyPersonType, DataParameter>(val => new DataParameter { Value = val.ToString(), DataType = DataType.VarChar });
    }
}
```

But if I use static MappingSchema.Default, then it is working as expected..

so when I use simple enum, it is enough to do:
```cs
MappingSchema.Default.SetConverter<string, MyPersonType>(s => Enum.Parse<MyPersonType>(s));
MappingSchema.Default.SetConverter<MyPersonType, DataParameter>(val => new DataParameter { Value = val.ToString(), DataType = DataType.VarChar });
```
and there is no need to apply DataType attribute on entity property

it's always frustrating when you can do something by many different ways, and some of them working some not and some are not obvious))

**Comment by sdanyliv** [comment]
> Why DataConnection->MappingSchema->SetConverter isn't working?

Well it is original design mistake. You have to do not change `MappingSchema` if you already use it. Even worse, never change `MappingSchema` of the `DataConnection`.

Create new `MappingSchema` and call `DataConnection.AddMappingSchema` if you need to do that dynamically.

Use `MappingSchema.Default` only if you are sure that these changes should be applied to all queries.

**Comment by feeleen** [comment]
@sdanyliv 

and what about fluent builder?

can I use one from `DataConnection` via `var builder = MappingSchema.GetFluentMappingBuilder();` (it woked fine for my case when did this `builder .Entity<Person>().Property(e => e.PersonType).HasConversion(v => v.Value, s => PersonType.Parse(s))`)

or I should get it only from new assigned `MappingSchema`? 

or `MappingSchema.Default` will work fine as well?

**Comment by dotarj** [comment]
> Currently there is no enum-specific fluent mapping API, but this should work:
> 
> ```cs
> mb.HasAttribute(typeof(SomeEnum).GetField(nameof(SomeEnum.Value1)), new MapValueAttribute(100));
> mb.HasAttribute(typeof(SomeEnum).GetField(nameof(SomeEnum.Value2)), new MapValueAttribute(200));
> ```

Hey @MaceWindu, I tried mapping enum values as described above (seemed like an easier way to map enums compared to the converter strategy), but it seems the method `EntityMappingBuilder<TEntity>.HasAttribute` now only support attributes inheriting from `MappingAttribute` and `MapValueAttribute` does not inherit from `MappingAttribute`. Is this done on purpose?

**Comment by MaceWindu** [comment]
@dotarj see #3825. Will fix in next release

**Comment by dotarj** [comment]
Wow, thanks for the amazingly fast response! And sorry for not checking out the issues before asking.

**Comment by MaceWindu** [comment]
Not at all. Completely forgot about it and you reminded me to fix it
